clients/src/main/java/org/apache/kafka/common/replica/RackAwareReplicaSelector.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.replica;
+
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Returns a replica whose rack id is equal to the rack id specified in the client request metadata. If no such replica
+ * is found, returns the leader.
+ */
+public class RackAwareReplicaSelector implements ReplicaSelector {
+
+    @Override
+    public Optional<ReplicaView> select(TopicPartition topicPartition,
+                                        ClientMetadata clientMetadata,
+                                        PartitionView partitionView) {
+        if (clientMetadata.rackId() != null && !clientMetadata.rackId().isEmpty()) {
+            Set<ReplicaView> sameRackReplicas = partitionView.replicas().stream()
+                    .filter(replicaInfo -> clientMetadata.rackId().equalsIgnoreCase(replicaInfo.endpoint().rack()))
```

Comment:
Also, why is this in the `client` project? I was under the impression that we wanted to make selection decision on the server.

Reply:
Thanks, that's a good catch. I'll preserve the case in the comparison. As for the `client` project, that's a good question. Since this is server impl code, it could also go in the core project, but then we'd need to rewrite it in Scala. @hachikuji wdyt?

Reply:
I think we don't want users depending on kafka-core. If we had a separate package for only server plugins, that might be another approach, but we have traditionally put common client and server code in kafka-clients. As for scala/java, we tend to prefer exposed interfaces to be written in Java since since that is easier for most users to work with. Using scala from java can be a bit of a bear, for example, but the opposite is usually straightforward.